http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/internal/RepeatableInputStreamRequestEntityTest.java

```diff
@@ -0,0 +1,799 @@
+                                                       .contentStreamProvider(provider)
+                                                       .build();
+
+        entity = new RepeatableInputStreamRequestEntity(request);
+
+        // First successful write
+        ByteArrayOutputStream firstOutput = new ByteArrayOutputStream();
+        entity.writeTo(firstOutput);
+        assertEquals(content, firstOutput.toString());
+
+        // Second write where reset should fail
+        IOException thrown = assertThrows(IOException.class,
+                                          () -> entity.writeTo(new ByteArrayOutputStream()));
+
+
+        assertEquals("Reset failed", thrown.getMessage());
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"100", "0", "9223372036854775807"}) // Long.MAX_VALUE
+    @DisplayName("parseContentLength should handle valid numeric values")
+    void parseContentLength_ValidNumbers_ParsesCorrectly(String contentLength) {
+        // Given
+        SdkHttpRequest httpRequest = httpRequestBuilder
+            .putHeader("Content-Length", contentLength)
+            .build();
+        HttpExecuteRequest request = HttpExecuteRequest.builder()
+                                                       .request(httpRequest)
+                                                       .build();
+
+        entity = new RepeatableInputStreamRequestEntity(request);
+        assertEquals(Long.parseLong(contentLength), entity.getContentLength());
+    }
+
+    @Test
+    @DisplayName("Multiple writes should work correctly with repeatable stream")
+    void writeTo_MultipleWrites_AllSucceed() throws IOException {
+        // Given
+        String content = "repeatable content";
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(content.getBytes());
+        ContentStreamProvider provider = () -> inputStream;
+
+        SdkHttpRequest httpRequest = httpRequestBuilder.build();
+        HttpExecuteRequest request = HttpExecuteRequest.builder()
+                                                       .request(httpRequest)
+                                                       .contentStreamProvider(provider)
+                                                       .build();
+
+        entity = new RepeatableInputStreamRequestEntity(request);
+
+        // Multiple writes
+        ByteArrayOutputStream output1 = new ByteArrayOutputStream();
+        ByteArrayOutputStream output2 = new ByteArrayOutputStream();
+        ByteArrayOutputStream output3 = new ByteArrayOutputStream();
+
+        entity.writeTo(output1);
+        entity.writeTo(output2);
+        entity.writeTo(output3);
+
+        // All outputs should contain the same content
+        assertEquals(content, output1.toString());
+        assertEquals(content, output2.toString());
+        assertEquals(content, output3.toString());
+    }
+
+    @Test
+    @DisplayName("Entity should handle mixed headers correctly")
```

Comment:
nit: should we name this "Entity should handle **_multiple_** headers correctly" ? since the headers are unique?